File: collie/src/main/java/com/snail/collie/Collie.java
package com.snail.collie;

import android.app.Activity;
import android.app.Application;
import android.os.Bundle;
import android.os.Handler;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.snail.collie.battery.BatteryInfo;
import com.snail.collie.battery.BatteryStatsTracker;
import com.snail.collie.debug.DebugHelper;
import com.snail.collie.fps.FpsTracker;
import com.snail.collie.mem.MemoryLeakTrack;
import com.snail.kotlin.mem.TrackMemoryInfo;
import com.snail.kotlin.CollieListener;
import com.snail.kotlin.Config;
import com.snail.kotlin.core.ActivityStack;
import com.snail.kotlin.core.CollieHandlerThread;
import com.snail.kotlin.fps.ITrackFpsListener;
import com.snail.kotlin.startup.LauncherTracker;
import com.snail.kotlin.trafficstats.ITrackTrafficStatsListener;
import com.snail.kotlin.trafficstats.TrafficStatsTracker;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;

public class Collie {

    private static volatile Collie sInstance = null;
    private Handler mHandler;
    private ITrackFpsListener mITrackListener;
    private MemoryLeakTrack.ITrackMemoryListener mITrackMemoryLeakListener;
    private BatteryStatsTracker.IBatteryListener mIBatteryListener;

    private List<CollieListener> mCollieListeners = new ArrayList<>();
    private HashSet<Application.ActivityLifecycleCallbacks> mActivityLifecycleCallbacks = new HashSet<>();

    private Collie() {
        mHandler = new Handler(CollieHandlerThread.INSTANCE.getLooper());
        mITrackListener = new ITrackFpsListener() {
            @Override
            public void onFpsTrack(final Activity activity, final long currentCostMils, final long currentDropFrame, final boolean isInFrameDraw, final long averageFps) {
                final long currentFps = currentCostMils == 0 ? 60 : Math.min(60, 1000 / currentCostMils);
                mHandler.post(new Runnable() {
                    @Override
                    public void run() {

                        if (currentDropFrame > 1)
                            DebugHelper.getInstance().update("实时fps " + currentFps +
                                    "\n 丢帧 " + currentDropFrame + " \n1s平均fps " + averageFps
                                    + " \n本次耗时 " + currentCostMils);

                        for (CollieListener collieListener : mCollieListeners) {
                            collieListener.onFpsTrack(activity, currentCostMils, currentDropFrame, isInFrameDraw, averageFps);
                        }
                    }

                });
            }

            @Override
            public void onANRAppear(Activity activity) {
                for (CollieListener collieListener : mCollieListeners) {
                    collieListener.onANRAppear(activity);
                }
            }
        };


        mITrackMemoryLeakListener = new MemoryLeakTrack.ITrackMemoryListener() {
            @Override
            public void onLeakActivity(String activity, int count) {
//                Log.v("Collie", "内存泄露 " + activity + " 数量 " + count);
                for (CollieListener collieListener : mCollieListeners) {
                    collieListener.onLeakActivity(activity, count);
                }
            }

            @Override
            public void onCurrentMemoryCost(TrackMemoryInfo trackMemoryInfo) {
//                Log.v("Collie", "内存  " + trackMemoryInfo.procName + " java内存  "
//                        + trackMemoryInfo.appMemory.dalvikPss + " native内存  " +
//                        trackMemoryInfo.appMemory.nativePss);
                for (CollieListener collieListener : mCollieListeners) {
                    collieListener.onCurrentMemoryCost(trackMemoryInfo);
                }
            }
        };

        mIBatteryListener = new BatteryStatsTracker.IBatteryListener() {
            @Override
            public void onBatteryCost(BatteryInfo batteryInfo) {
                for (CollieListener collieListener : mCollieListeners) {
                    collieListener.onBatteryCost(batteryInfo);
                }
            }
        };
    }

    public static Collie getInstance() {
        if (sInstance == null) {
            synchronized (Collie.class) {
                if (sInstance == null) {
                    sInstance = new Collie();
                }
            }
        }
        return sInstance;
    }

    public void addActivityLifecycleCallbacks(Application.ActivityLifecycleCallbacks callbacks) {
        mActivityLifecycleCallbacks.add(callbacks);
    }

    public void removeActivityLifecycleCallbacks(Application.ActivityLifecycleCallbacks callbacks) {
        mActivityLifecycleCallbacks.remove(callbacks);
    }

    private Application.ActivityLifecycleCallbacks mActivityLifecycleCallback = new Application.ActivityLifecycleCallbacks() {
        @Override
        public void onActivityCreated(@NonNull Activity activity, @Nullable Bundle bundle) {
            ActivityStack.INSTANCE.push(activity);
            for (Application.ActivityLifecycleCallbacks item : mActivityLifecycleCallbacks) {
                item.onActivityCreated(activity, bundle);
            }
        }

        @Override
        public void onActivityStarted(@NonNull final Activity activity) {
            ActivityStack.INSTANCE.markStart();
            for (Application.ActivityLifecycleCallbacks item : mActivityLifecycleCallbacks) {
                item.onActivityStarted(activity);
            }
        }

        @Override
        public void onActivityResumed(@NonNull Activity activity) {
            for (Application.ActivityLifecycleCallbacks item : mActivityLifecycleCallbacks) {
                item.onActivityResumed(activity);
            }
        }

        @Override
        public void onActivityPaused(@NonNull Activity activity) {
            for (Application.ActivityLifecycleCallbacks item : mActivityLifecycleCallbacks) {
                item.onActivityPaused(activity);
            }
        }

        @Override
        public void onActivityStopped(@NonNull Activity activity) {
            ActivityStack.INSTANCE.markStop();
            for (Application.ActivityLifecycleCallbacks item : mActivityLifecycleCallbacks) {
                item.onActivityStopped(activity);
            }
        }

        @Override
        public void onActivitySaveInstanceState(@NonNull Activity activity, @NonNull Bundle bundle) {
            for (Application.ActivityLifecycleCallbacks item : mActivityLifecycleCallbacks) {
                item.onActivitySaveInstanceState(activity, bundle);
            }
        }

        @Override
        public void onActivityDestroyed(@NonNull Activity activity) {
            for (Application.ActivityLifecycleCallbacks item : mActivityLifecycleCallbacks) {
                item.onActivityDestroyed(activity);
            }
            ActivityStack.INSTANCE.pop(activity);
        }
    };

    public void init(@NonNull Application application,
                     final Config config,
                     final CollieListener listener) {
        application.registerActivityLifecycleCallbacks(mActivityLifecycleCallback);
        mCollieListeners.add(listener);

        if (config.getUseTrafficTrack()) {
            TrafficStatsTracker.INSTANCE.setTrafficStatsListener(new ITrackTrafficStatsListener() {
                @Override
                public void onTrafficStats(Activity activity, long value) {
                    for (CollieListener collieListener : mCollieListeners) {
                        collieListener.onTrafficStats(activity, value);
                    }
                }
            });
            TrafficStatsTracker.INSTANCE.startTrack(application);
        }
        if (config.getUseMemTrack()) {
            MemoryLeakTrack.getInstance().startTrack(application);
            MemoryLeakTrack.getInstance().addOnMemoryLeakListener(mITrackMemoryLeakListener);
        }
        if (config.getUseFpsTrack()) {
            FpsTracker.getInstance().setTrackerListener(mITrackListener);
            FpsTracker.getInstance().startTrack(application);
        }
        if (config.getShowDebugView()) {
            DebugHelper.getInstance().startTrack(application);
        }

        if (config.getUseBatteryTrack()) {
            BatteryStatsTracker.getInstance().addBatteryListener(mIBatteryListener);
            BatteryStatsTracker.getInstance().startTrack(application);
        }

        if (config.getUseStartUpTrack()) {
            LauncherTracker.INSTANCE.setILaunchTrackListener(new LauncherTracker.ILaunchTrackListener() {
                @Override
                public void onActivityFocusableCost(@Nullable Activity activity, long duration, boolean finishNow) {
                    for (CollieListener collieListener : mCollieListeners) {
                        collieListener.onActivityFocusableCost(activity, duration, finishNow);
                    }
                }

                @Override
                public void onAppColdLaunchCost(long duration, String processName) {
                    for (CollieListener collieListener : mCollieListeners) {
                        collieListener.onAppColdLaunchCost(duration, processName);
                    }
                }

                @Override
                public void onActivityLaunchCost(Activity activity, long duration, boolean finishNow) {
                    for (CollieListener collieListener : mCollieListeners) {
                        collieListener.onActivityLaunchCost(activity, duration, finishNow);
                    }
                }
            });

            LauncherTracker.INSTANCE.startTrack(application);
        }

    }

    public void registerCollieListener(CollieListener listener) {
        mCollieListeners.add(listener);
    }

    public void unRegisterCollieListener(CollieListener listener) {
        mCollieListeners.remove(listener);
    }

    public void stop(@NonNull Application application) {
        application.unregisterActivityLifecycleCallbacks(mActivityLifecycleCallback);
        CollieHandlerThread.INSTANCE.quitSafely();
    }
}

File: collie/src/main/java/com/snail/kotlin/Collie.kt
package com.snail.kotlin

import android.app.Activity
import android.app.Application
import android.os.Bundle
import android.os.Handler
import com.snail.kotlin.core.ActivityStack.push
import com.snail.kotlin.core.ActivityStack.markStart
import com.snail.kotlin.core.ActivityStack.markStop
import com.snail.kotlin.core.ActivityStack.pop
import com.snail.kotlin.trafficstats.TrafficStatsTracker.trafficStatsListener
import com.snail.kotlin.startup.LauncherTracker.iLaunchTrackListener
import com.snail.kotlin.fps.ITrackFpsListener
import com.snail.kotlin.mem.MemoryLeakTrack.ITrackMemoryListener
import com.snail.kotlin.battery.BatteryStatsTracker.IBatteryListener
import com.snail.kotlin.trafficstats.TrafficStatsTracker
import com.snail.kotlin.trafficstats.ITrackTrafficStatsListener
import com.snail.kotlin.mem.MemoryLeakTrack
import com.snail.kotlin.fps.FpsTracker
import com.snail.kotlin.debug.DebugHelper
import com.snail.kotlin.battery.BatteryStatsTracker
import com.snail.kotlin.startup.LauncherTracker
import com.snail.kotlin.startup.LauncherTracker.ILaunchTrackListener
import com.snail.kotlin.core.CollieHandlerThread
import kotlin.jvm.Volatile
import com.snail.kotlin.battery.BatteryInfo
import com.snail.kotlin.mem.TrackMemoryInfo
import java.util.ArrayList
import java.util.HashSet

class Collie private constructor() {
    private val mHandler: Handler = Handler(CollieHandlerThread.looper)
    private val mITrackListener: ITrackFpsListener
    private val mITrackMemoryLeakListener: ITrackMemoryListener
    private val mIBatteryListener: IBatteryListener
    private val mCollieListeners: MutableList<CollieListener> = ArrayList()
    private val mActivityLifecycleCallbacks = HashSet<Application.ActivityLifecycleCallbacks>()
    fun addActivityLifecycleCallbacks(callbacks: Application.ActivityLifecycleCallbacks) {
        mActivityLifecycleCallbacks.add(callbacks)
    }

    fun removeActivityLifecycleCallbacks(callbacks: Application.ActivityLifecycleCallbacks) {
        mActivityLifecycleCallbacks.remove(callbacks)
    }

    private val mActivityLifecycleCallback: Application.ActivityLifecycleCallbacks =
        object : Application.ActivityLifecycleCallbacks {
            override fun onActivityCreated(activity: Activity, bundle: Bundle?) {
                push(activity)
                for (item in mActivityLifecycleCallbacks) {
                    item.onActivityCreated(activity, bundle)
                }
            }

            override fun onActivityStarted(activity: Activity) {
                markStart()
                for (item in mActivityLifecycleCallbacks) {
                    item.onActivityStarted(activity)
                }
            }

            override fun onActivityResumed(activity: Activity) {
                for (item in mActivityLifecycleCallbacks) {
                    item.onActivityResumed(activity)
                }
            }

            override fun onActivityPaused(activity: Activity) {
                for (item in mActivityLifecycleCallbacks) {
                    item.onActivityPaused(activity)
                }
            }

            override fun onActivityStopped(activity: Activity) {
                markStop()
                for (item in mActivityLifecycleCallbacks) {
                    item.onActivityStopped(activity)
                }
            }

            override fun onActivitySaveInstanceState(activity: Activity, bundle: Bundle) {
                for (item in mActivityLifecycleCallbacks) {
                    item.onActivitySaveInstanceState(activity, bundle)
                }
            }

            override fun onActivityDestroyed(activity: Activity) {
                for (item in mActivityLifecycleCallbacks) {
                    item.onActivityDestroyed(activity)
                }
                pop(activity)
            }
        }

    fun init(
        application: Application,
        config: Config,
        listener: CollieListener
    ) {
        application.registerActivityLifecycleCallbacks(mActivityLifecycleCallback)
        mCollieListeners.add(listener)
        if (config.useTrafficTrack) {
            trafficStatsListener = object : ITrackTrafficStatsListener {
                override fun onTrafficStats(activity: Activity, value: Long) {
                    for (collieListener in mCollieListeners) {
                        collieListener.onTrafficStats(activity, value)
                    }
                }
            }
            TrafficStatsTracker.startTrack(application)
        }
        if (config.useMemTrack) {
            MemoryLeakTrack.instance!!.startTrack(application)
            MemoryLeakTrack.instance!!.addOnMemoryLeakListener(mITrackMemoryLeakListener)
        }
        if (config.useFpsTrack) {
            FpsTracker.getInstance().setTrackerListener(mITrackListener)
            FpsTracker.getInstance().startTrack(application)
        }
        if (config.showDebugView) {
            DebugHelper.instance!!.startTrack(application)
        }
        if (config.useBatteryTrack) {
            BatteryStatsTracker.instance!!.addBatteryListener(mIBatteryListener)
            BatteryStatsTracker.instance!!.startTrack(application)
        }
        if (config.useStartUpTrack) {
            iLaunchTrackListener = object : ILaunchTrackListener {
                override fun onActivityFocusableCost(
                    activity: Activity?,
                    duration: Long,
                    finishNow: Boolean
                ) {
                    for (collieListener in mCollieListeners) {
                        collieListener.onActivityFocusableCost(activity, duration, finishNow)
                    }
                }

                override fun onAppColdLaunchCost(duration: Long, processName: String?) {
                    for (collieListener in mCollieListeners) {
                        collieListener.onAppColdLaunchCost(duration, processName)
                    }
                }

                override fun onActivityLaunchCost(
                    activity: Activity?,
                    duration: Long,
                    finishNow: Boolean
                ) {
                    for (collieListener in mCollieListeners) {
                        collieListener.onActivityLaunchCost(activity, duration, finishNow)
                    }
                }
            }
            LauncherTracker.startTrack(application)
        }
    }

    fun stop(application: Application) {
        application.unregisterActivityLifecycleCallbacks(mActivityLifecycleCallback)
        CollieHandlerThread.quitSafely()
    }

    companion object {
        @Volatile
        private var sInstance: Collie? = null

        @JvmStatic
        val instance: Collie
            get() {
                if (sInstance == null) {
                    synchronized(Collie::class.java) {
                        if (sInstance == null) {
                            sInstance = Collie()
                        }
                    }
                }
                return sInstance!!
            }
    }

    init {
        mITrackListener = object : ITrackFpsListener {
            override fun onFpsTrack(
                activity: Activity,
                currentCostMils: Long,
                currentDropFrame: Long,
                isInFrameDraw: Boolean,
                averageFps: Long
            ) {
                val currentFps =
                    if (currentCostMils == 0L) 60 else Math.min(60, 1000 / currentCostMils)
                mHandler.post {
                    if (currentDropFrame > 1) DebugHelper.instance!!.update(
                        """实时fps $currentFps
 丢帧 $currentDropFrame 
1s平均fps $averageFps 
本次耗时 $currentCostMils"""
                    )
                    for (collieListener in mCollieListeners) {
                        collieListener.onFpsTrack(
                            activity,
                            currentCostMils,
                            currentDropFrame,
                            isInFrameDraw,
                            averageFps
                        )
                    }
                }
            }

            override fun onANRAppear(activity: Activity?) {
                for (collieListener in mCollieListeners) {
                    collieListener.onANRAppear(activity)
                }
            }
        }
        mITrackMemoryLeakListener = object : ITrackMemoryListener {
            override fun onLeakActivity(activity: String?, count: Int) {
//                Log.v("Collie", "内存泄露 " + activity + " 数量 " + count);
                for (collieListener in mCollieListeners) {
                    collieListener.onLeakActivity(activity, count)
                }
            }

            override fun onCurrentMemoryCost(trackMemoryInfo: TrackMemoryInfo?) {
//                Log.v("Collie", "内存  " + trackMemoryInfo.procName + " java内存  "
//                        + trackMemoryInfo.appMemory.dalvikPss + " native内存  " +
//                        trackMemoryInfo.appMemory.nativePss);
                for (collieListener in mCollieListeners) {
                    collieListener.onCurrentMemoryCost(trackMemoryInfo)
                }
            }
        }
        mIBatteryListener = object : IBatteryListener {
            override fun onBatteryCost(batteryInfo: BatteryInfo?) {
                for (collieListener in mCollieListeners) {
                    collieListener.onBatteryCost(batteryInfo)
                }
            }
        }
    }
}